### PR TITLE
add mjpeg stream helper class to basedriver

### PR DESCRIPTION
**Pull request by jlipps**
There is this "standard" of sorts called [MJPEG](https://en.wikipedia.org/wiki/Motion_JPEG), which allows for streaming of images over HTTP. Some users of Appium have MJPEG servers connected to the devices they use with Appium, which deliver relatively high-framerate images off the device. If this is the case, why not just use it for screenshots as well and bypass the sometimes time-consuming Appium-internal screenshot commands?

MJPEG is a sort-of "push" mechanism, so in this PR I just provide a class which maintains a reference to the last-pushed image, which would then be available to any driver that wants it. I'll start the driver PRs soon, but the basic idea is that this will be behind a capability, which if turned on will just send back the last-received image immediately upon screenshot request. Something like this:

```js
// in driver session start
if (this.opts.mjpegScreenshotUrl) {
  this.mjpegStream = new MJpegStream(this.opts.mjpegScreenshotUrl);
  this.mjpegStream.start();
}
```

```js
// in screenshot command implementation 
if (this.mjpegStream && this.mjpegStream.lastChunkBase64) {
  return this.mjpegStream.lastChunkBase64;
}

// now continue with regular screenshot logic as a fallback
```

```js
// in driver session stop/cleanup
if (this.mjpegStream) {
  this.mjpegStream.stop();
}
```

### CODE REVIEW

File: lib/basedriver/mjpeg.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    this.url = mJpegUrl;
+    this.clear();
+  }
+
+  /**
+   * Get the base64-encoded version of the JPEG
+   * @returns {string}
+   */
+  get lastChunkBase64 () {
+    if (!this.lastChunk) {
```
Comment (mykola-mokhnach) [inline lib/basedriver/mjpeg.js]: Ternary operator might be shorter here

File: lib/basedriver/mjpeg.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    this.request = null;
+    this.consumer = null;
+    this.lastChunk = null;
+    this.updateCount = 0;
+  }
+
+  /**
+   * Start reading the MJpeg stream and storing the last image
+   */
+  start () {
```
Comment (mykola-mokhnach) [inline lib/basedriver/mjpeg.js]: why not async?

File: lib/basedriver/mjpeg.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    this.request
+      .on('error', onErr)   // ensure we do something with errors
+      .pipe(this.consumer)  // allow chunking and transforming of jpeg data
+      .pipe(this);          // send the actual jpegs to ourself
+  }
+
+  /**
+   * Stop reading the MJpeg stream. Ensure we disconnect all the pipes and stop
+   * the HTTP request itself. Then reset the state.
```
Comment (mykola-mokhnach) [inline lib/basedriver/mjpeg.js]: It might be useful to mark overridden method with `@Override`

File: test/basedriver/fixtures/mjpeg-server.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+export function initMJpegServer (port, intMs = 300, times = 20) {
+  const server = http.createServer(async function (req, res) {
+    const mJpegReqHandler = mJpegServer.createReqHandler(req, res);
+    const jpg = await fs.readFile(IMAGE_PATH, 'binary');
+
+    // just send the same jpeg over and over
+    for (let i = 0; i < times; i++) {
+      await B.delay(intMs);
+      mJpegReqHandler._write(jpg, null, () => {});
```
Comment (mykola-mokhnach) [inline test/basedriver/fixtures/mjpeg-server.js]: there is also _.noop helper

**Comment by mykola-mokhnach** [comment]
We probably would need some fast PNG->JPEG converter here as well

**Comment by mykola-mokhnach** [comment]
> For now my thought is if someone uses this, they know they will get jpeg formatted screenshots out of appium instead of png.

Selenium's getScreenshot always returns images in PNG format. There are no other endpoints, which allow to get JPEG directly. That is why I suppose we need to do the conversion on Appium side

File: lib/basedriver/mjpeg.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    this.url = mJpegUrl;
+    this.clear();
+  }
+
+  /**
+   * Get the base64-encoded version of the JPEG
+   * @returns {string}
+   */
+  get lastChunkBase64 () {
+    return this.lastChunk ? this.lastChunk.toString('base64') : null;
```
Comment (mykola-mokhnach) [inline lib/basedriver/mjpeg.js]: _.isBuffer(lastChunk) ?

File: lib/basedriver/mjpeg.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    this.request = null;
+    this.consumer = null;
+    this.lastChunk = null;
+    this.updateCount = 0;
+  }
+
+  /**
+   * Start reading the MJpeg stream and storing the last image
+   */
+  async start () {
```
Comment (mykola-mokhnach) [inline lib/basedriver/mjpeg.js]: 👍

File: lib/basedriver/mjpeg.js [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+    await startPromise;
+  }
+
+  /**
+   * Stop reading the MJpeg stream. Ensure we disconnect all the pipes and stop
+   * the HTTP request itself. Then reset the state.
+   */
+  stop () {
+    if (this.consumer) {
```
Comment (mykola-mokhnach) [inline lib/basedriver/mjpeg.js]: if (!this.consumer) return

File: lib/basedriver/mjpeg.js [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    };
+
+    this.request = request(this.url);
+
+    this.request
+      .on('error', onErr)   // ensure we do something with errors
+      .pipe(this.consumer)  // allow chunking and transforming of jpeg data
+      .pipe(this);          // send the actual jpegs to ourself
+
+    await startPromise;
```
[the remark this answers is not in the document]
Comment (jlipps) [reply lib/basedriver/mjpeg.js]: there is no last image, so that's not possible.

[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply lib/basedriver/mjpeg.js]: I mean this might block forever if the server does not produce any data

File: lib/basedriver/mjpeg.js (hunk 3 above)
Comment (jlipps) [reply lib/basedriver/mjpeg.js]: this method is not overridden

File: lib/basedriver/mjpeg.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (jlipps) [reply lib/basedriver/mjpeg.js]: oh right. that would be an error. probably we should have a timeout and if the timeout gets hit, throw an error

File: lib/basedriver/mjpeg.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+  get lastChunkBase64 () {
+    return this.lastChunk ? this.lastChunk.toString('base64') : null;
+  }
+
+
+  /**
+   * Reset internal state
+   */
+  clear () {
+    this.didStart = false;
```
[the remark this answers is not in the document]
Comment (jlipps) [reply lib/basedriver/mjpeg.js]: nope! will remove

**Comment by jlipps** [comment]
@mykola-mokhnach @imurchie what do you think, should this be a part of appium-support instead? now that it is just a class which can be used, but not tied to the driver code itself.

**Comment by mykola-mokhnach** [comment]
If the class has no direct dependencies on base driver then yes, it would be better to keep it in appium-support

File: lib/basedriver/mjpeg.js [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    // use the deferred pattern so we can wait for the start of the stream
+    // based on what comes in from an external pipe
+    let startPromise = new B((res, rej) => {
+      this.registerStartSuccess = res;
+      this.registerStartFailure = rej;
+    });
+
+    // start a timeout so that if the server does not return data, we don't
+    // block forever.
+    setTimeout(() => {
```
Comment (mykola-mokhnach) [inline lib/basedriver/mjpeg.js]: I think bluebird promises already have the standard .timeout method

**Comment by jlipps** [comment]
closing, gonna move to appium-support

**Pull request closed without merging**